[camerax] Implement video capture for CameraX android camera re-write

gmackall opened the pull request:
This PR implements video capture for the CameraX re-write of the android camera plugin.

https://github.com/flutter/flutter/issues/111138

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

gmackall commented on packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:
```diff
@@ -379,4 +397,41 @@ class AndroidCameraCameraX extends CameraPlatform {
+    recorder = Recorder(); // TODO(gmackall): Configure resolution info here.
+    videoCapture = await VideoCapture.withOutput(recorder!);
+    processCameraProvider!.bindToLifecycle(cameraSelector!, [videoCapture!]);
+    videoOutputPath = await SystemServices.getTempFilePath('MOV', '.mp4');
```
The main reason I wrapped it was because the video capture classes don't return the path at any point (only take it in a constructor). 

So given that it had to be returned from the dart side, it felt easiest to wrap the initial creation of the path on the dart side and then store till returning.

camsim99 replied:
That makes sense! We can modify image capture to use the wrapped method in the future. I can file an issue

reidbaker replied:
Consider making "MOV" A named constant. 
Can you link to the documentation that says we will get an mp4 here or if the file extension does not matter pick something that does not imply a format like "temp"

gmackall replied:
From [this page](https://developer.android.com/training/camerax/video-capture#bind-videocapture), the Recorder "selects the most suitable format for the system," so it looks like we are not guaranteed an mp4. Will change to temp.

gmackall commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/RecorderHostApiImpl.java:
```diff
@@ -0,0 +1,102 @@
+  public void create(@NonNull Long instanceId, Long aspectRatio, Long bitRate) {
+    Recorder.Builder recorderBuilder = cameraXProxy.createRecorderBuilder();
+    if (aspectRatio != null) {
+      recorderBuilder.setAspectRatio(Math.toIntExact(aspectRatio));
```
I'm open to either, but it brings up a sort of general question I had. In wrapping these CameraX classes, should we strive to expose all the functionality of the original class (even when not needed), or should we only be exposing exactly what we are using? 

I had a general impression that wrapping as closely to the original class (when easy to do so) was generally better, which was why I left it in despite not using it. But again I am fine removing if you think that is better!

camsim99 replied:
We discussed this offline and determined it's good to leave as is.

In short, we are willing to compress/not wrap functionality that is unlikely to be needed by developers to modify the plugin or build their own with the wrapped classes. Otherwise, we can expose functionality as closely as possible for clarity.

reidbaker commented on packages/camera/camera_android_camerax/example/lib/main.dart:
```diff
@@ -519,16 +519,16 @@ class _CameraExampleHomeState extends State<CameraExampleHome>
-              cameraController == null ? null :
-              (cameraController.value.isRecordingPaused
-                  ? onResumeButtonPressed : onPauseButtonPressed),
+          onPressed: cameraController == null
```
Nested ternary operators can be difficult to read. Can you break this apart?

reidbaker commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PendingRecordingHostApiImpl.java:
```diff
@@ -0,0 +1,72 @@
+    recordingFlutterApi = new RecordingFlutterApiImpl(binaryMessenger, instanceManager);
+  }
+
+  public void setContext(Context context) {
```
Is there guidance about if this is an application context or an activity context? My guess is activity.

reidbaker commented on packages/camera/camera_android_camerax/lib/src/recorder.dart:
```diff
@@ -0,0 +1,122 @@
+import 'pending_recording.dart';
+
+/// A dart wrapping of the CameraX Recorder class. Does not exactly wrap all
+/// methods.
```
+1

reidbaker commented on packages/camera/camera_android_camerax/lib/src/system_services.dart:
```diff
@@ -61,6 +61,14 @@ class SystemServices {
     api.stopListeningForDeviceOrientationChange();
   }
+
+  /// Gets the current cache directory path from context.
```
Can you expand a bit on this documentation? For example what would change the current cache directory, who else uses the current cache directory, are there constraints on its use or the files that can be put in it.

camsim99 commented on packages/camera/camera_android_camerax/lib/src/system_services.dart:
```diff
@@ -61,6 +61,17 @@ class SystemServices {
     api.stopListeningForDeviceOrientationChange();
   }
+
+  /// Gets a file path to use to create a temp file at. It is created by calling
+  /// File.createTempFile(prefix, suffix, cacheDir), where cacheDir is the
+  /// cache directory identified by the current application context
+  /// using context.getCacheDir().
```
Nit:
```suggestion
  /// Gets a file path to use to create a temp file at. It is created by calling
  /// File.createTempFile(prefix, suffix, cacheDir) on the native side, where
  ///  cacheDir is the cache directory identified by the current application
  /// context using context.getCacheDir().
```

camsim99 commented on packages/camera/camera_android_camerax/lib/src/video_capture.dart:
```diff
@@ -0,0 +1,116 @@
+          binaryMessenger: binaryMessenger, instanceManager: instanceManager);
+    });
+    final int videoCaptureId = await withOutput(identifier);
+    return instanceManager.getInstanceWithWeakReference(videoCaptureId)!
+        as VideoCapture;
```
I have received lint errors by using this instead of:
```suggestion
    return instanceManager.getInstanceWithWeakReference<VideoCapture>(videoCaptureId)!;
```

camsim99 commented on packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart:
```diff
@@ -388,6 +396,44 @@ void main() {
+    verify(recording.close());
+    verifyNoMoreInteractions(recording);
+  });
+
```
Can you also add a test for `startVideoRecording`?

camsim99 commented on packages/camera/camera_android_camerax/test/pending_recording_test.dart:
```diff
@@ -0,0 +1,48 @@
+    when(mockApi.start(pendingRecordingId)).thenReturn(mockRecordingId);
+    expect(await pendingRecording.start(), mockRecording);
+    verify(mockApi.start(pendingRecordingId));
+  });
```
Can you also add a test for the Flutter API create method?

reidbaker replied:
@camsim99 can you resolve this if you are happy with the test added?

camsim99 replied:
Nit: can you make the test title more specific? Otherwise, lgtm

camsim99 commented on packages/camera/camera_android_camerax/test/recording_test.dart:
```diff
@@ -0,0 +1,104 @@
+
+      verify(mockApi.stop(recordingId));
+    });
+  });
```
Can you add a test for the Flutter API create method?

camsim99 commented on packages/camera/camera_android_camerax/test/video_capture_test.dart:
```diff
@@ -0,0 +1,74 @@
+    when(mockApi.getOutput(videoCaptureId)).thenReturn(mockRecorderId);
+    expect(await videoCapture.getOutput(), mockRecorder);
+    verify(mockApi.getOutput(videoCaptureId));
+  });
```
Can you add a test for the Flutter API create method?

camsim99 reviewed (commented):
Looking good! I left a comment on [one main concern](https://github.com/flutter/packages/pull/3467#discussion_r1172860624) about using aspect ratio instead of resolution (I think we actually can use resolution), but otherwise, just requests for some tests and nits.

gmackall commented on packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:
```diff
@@ -39,8 +43,36 @@ class AndroidCameraCameraX extends CameraPlatform {
+  /// Whether or not the [preview] is currently bound to the lifecycle that the
+  /// [processCameraProvider] tracks.
+  @visibleForTesting
+  bool previewIsBound = false;
```
Hmm I must have messed up a merge somewhere, I'll fix so that this stays removed

gmackall commented on packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:
```diff
@@ -419,6 +451,46 @@ class AndroidCameraCameraX extends CameraPlatform {
+  /// Stops the video recording and returns the file where it was saved.
+  @override
+  Future<XFile> stopVideoRecording(int cameraId) async {
+    recording!.close();
```
Hmm it looks like in the native side we return success but with null path, and then on the [dart side we throw an error](https://github.com/flutter/packages/blob/main/packages/camera/camera_android/lib/src/android_camera.dart#L280)? Will look into this because I'm not sure what the best behavior is given that the interface makes us return a non-null file.

gmackall commented on packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PendingRecordingTest.java:
```diff
@@ -0,0 +1,71 @@
+    final Long mockRecordingId = testInstanceManager.addHostCreatedInstance(mockRecording);
+    testInstanceManager.addDartCreatedInstance(mockPendingRecording, mockPendingRecordingId);
+
+    doReturn(mockRecording).when(mockPendingRecording).start(any(), any());
```
I've made the handleVideoRecordEvent visibleForTesting and added a test, let me know if you think that tests it sufficiently!

camsim99 reviewed (approved):
LGTM!

reidbaker reviewed (commented):
partial review.

reidbaker commented:
before continuing my pass I need the cla to be figured out.

camsim99 commented on packages/camera/camera_android_camerax/lib/src/video_capture.dart:
```diff
@@ -0,0 +1,116 @@
+      {BinaryMessenger? binaryMessenger, InstanceManager? instanceManager})
+      : super.detached(
+            binaryMessenger: binaryMessenger,
+            instanceManager: instanceManager) {
```
I missed this -- we should do this here too!

gmackall replied:
The instance manager here in VideoCapture is only being used in the constructor of the VideoCaptureHostApiImpl, and in that constructor we [override to the global instance manager](https://github.com/flutter/packages/pull/3467/files#diff-ed895c82b184dbb68b28c6681dbb1884fd75cc3e02cece32901b5239e05c6e2aR52). This is the same pattern that is being used elsewhere, for examples see the [Preview constructor](https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/lib/src/preview.dart#L17) vs the [PreviewHostApiImpl constructor](https://github.com/flutter/packages/blob/main/packages/camera/camera_android_camerax/lib/src/preview.dart#L78).

reidbaker replied:
@camsim99 or @gmackall  can one of you add to your growing clean up bug to make this pattern consistent. If we cant keep it straight then neither can first time contributors.

gmackall replied:
I believe the pattern here is consistent. 

In the `Thing` constructor, we pass in an `InstanceManager? instanceManager`, and then use that to create a `ThingHostApiImpl`, and then in the `ThingHostApiImpl` constructor we check if that `instanceManager` we passed is null and override to the global if so.

gmackall replied:
Actually, I see that some of the newer wrapped classes like zoom state and exposure state do not create these HostApiImpls in their detached constructors. I'll file an issue to ensure the detached versions of the constructors all align on whether to create or not (probably not, given the point is that they are not attached to a native object and therefore do not need to call host api methods).

gmackall replied:
https://github.com/flutter/flutter/issues/126293

gmackall commented:
Heads up that I have created a new version of this pr with commits up to this point squashed to fix the CLA issue @reidbaker pointed out. Leaving this open temporarily while the outstanding comments are unresolved.

https://github.com/flutter/packages/pull/3899

gmackall commented on packages/camera/camera_android_camerax/lib/src/video_capture.dart:
```diff
@@ -0,0 +1,116 @@
+            instanceManager: instanceManager) {
+    _api = VideoCaptureHostApiImpl(
+        binaryMessenger: binaryMessenger, instanceManager: instanceManager);
+    AndroidCameraXCameraFlutterApis.instance.ensureSetUp();
```
This is currently the pattern being used in many other constructors, would you be okay with this being merged as is with an issue to fix all occurrences of this AndroidCameraXCameraFlutterApis setup?

camsim99 replied:
Can you file an issue for that so we can track this task?

gmackall replied:
Is this captured under the first point of https://github.com/flutter/flutter/issues/125926, or should I make it its own issue?

camsim99 replied:
Adding it there is perfectly fine!

gmackall commented on packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart:
```diff
@@ -471,6 +484,83 @@ void main() {
     expect(previewTexture.textureId, equals(textureId));
   });
 
+  test(
```
Made them share a group in the new PR

reidbaker commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/SystemServicesHostApiImpl.java:
```diff
@@ -108,4 +120,16 @@ public void stopListeningForDeviceOrientationChange() {
+  /** Returns a path to be used to create a temp file in the current cache directory. */
+  @Override
+  @NonNull
+  public String getTempFilePath(@NonNull String prefix, @NonNull String suffix) {
```
nit: filePrefux and fileSuffix, as opposed to path or some other definition. Also acceptable would be updating the documentation. 
you also might consider making these named parameters for call site readability.

reidbaker commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/SystemServicesHostApiImpl.java:
```diff
@@ -108,4 +120,16 @@ public void stopListeningForDeviceOrientationChange() {
+    try {
+      File path = File.createTempFile(prefix, suffix, context.getCacheDir());
+      return path.toString();
+    } catch (IOException | SecurityException e) {
```
If you know when these might be expected consider adding a comment.

reidbaker commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/SystemServicesHostApiImpl.java:
```diff
@@ -108,4 +120,16 @@ public void stopListeningForDeviceOrientationChange() {
+      File path = File.createTempFile(prefix, suffix, context.getCacheDir());
+      return path.toString();
+    } catch (IOException | SecurityException e) {
+      throw new RuntimeException(e);
```
Should this be a typed exception?

gmackall replied:
At the time it was written, it was only possible to use RuntimeException here (the general pattern was that this method is wrapped in a try catch anyways in GeneratedCameraXLibrary, and then GeneratedCameraXLibrary would throw a PlatformException).

Since updating to pigeon 9.10, we can throw FlutterError instead (there is some discussion about the old pattern vs new FlutterError [on this old PR](https://github.com/flutter/packages/pull/3561#event-8858727980)). I will change here to throw a FlutterError.

reidbaker commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/RecordingHostApiImpl.java:
```diff
@@ -0,0 +1,50 @@
+    this.instanceManager = instanceManager;
+  }
+
+  @Override
```
Nit: consider ordering these methods in the order they will likely be called or into related pairs.

reidbaker commented on packages/camera/camera_android_camerax/lib/src/system_services.dart:
```diff
@@ -61,6 +61,19 @@ class SystemServices {
+  /// File.createTempFile(prefix, suffix, cacheDir), on the android side, where
+  /// where cacheDir is the cache directory identified by the current application
+  /// context using context.getCacheDir().
+  static Future<String> getTempFilePath(String prefix, String suffix,
```
I think from the tests I read prefix is actually the file name and suffix is the extension. That came as a bit of a suprise from reading the documentation I would have expected `prefix<something the caller does not control>suffix` for the file name. 

For example it means as a caller I think this means I need to ensure that I pass something different each time  or risk overwriting the previous file.

gmackall replied:
I'll clarify the documentation before merging, the way it actually works is that the file ends up being
`prefix<something the caller does not control.suffix`, i.e. 'suffix' is the file extension, and prefix is _part_ of the file name, but more gets added by createTempFile (a random sequence of numbers).

So `getTempFile('MOV', '.temp')`, is something like `<cachedir>/MOV1231242244.temp`.

FWIW I chose prefix and suffix because those are the variable names used by the native methods implementation/documentation.

reidbaker replied:
OK in that case it means the documentation *is correct* but that is not what the tests are testing for.  I think using the same names as the native code is a reasonable choice. 

Specifically in packages/camera/camera_android_camerax/test/system_services_test.dart and in packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/SystemServicesTest.java you are testing for the result to be equal to`testPath + testPrefix + testSuffix` (or the file for prefix and suffix). Instead of testing that the output had a path of `testPath` a file prefix of `testPrefix` and an extension of `testSuffix`. Which is fine it is just confusing when trying to understand what to expect from the tests.

reidbaker reviewed (commented):
Final approval is up for @camsim99 to decide. I know there is a balance to quality and speed we are trying to hit.

camsim99 commented on packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PendingRecordingFlutterApiImpl.java:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  void create(@NonNull PendingRecording pendingRecording, @Nullable Reply<Void> reply) {
+    create(instanceManager.addHostCreatedInstance(pendingRecording), reply);
```
Added to https://github.com/flutter/flutter/issues/125926

camsim99 commented on packages/camera/camera_android_camerax/lib/src/recording.dart:
```diff
@@ -0,0 +1,120 @@
+
+  /// Closes the specified recording instance.
+  Future<void> closeFromInstance(Recording recording) async {
+    close(instanceManager.getIdentifier(recording)!);
```
I agree it may be a useful tech debt item to include a more helpful error message, but hitting this means most likely someone adding new wrapped classes made an error or maybe someone modifying the plugin implementation did. It won't crash the app, either, so leaving it like this should be ok for now.

camsim99 commented on packages/camera/camera_android_camerax/pigeons/camerax_library.dart:
```diff
@@ -161,6 +163,60 @@ abstract class PreviewHostApi {
 
+@HostApi(dartHostTestHandler: 'TestVideoCaptureHostApi')
+abstract class VideoCaptureHostApi {
+  int withOutput(int videoOutputId);
```
FWIW, he's following a pattern I started. We should consider copying the documentation from the wrapped classes to these objects in a future cleanup PR.

camsim99 commented on packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart:
This sort of testing is good, and likely the sort we should use for several other methods tested here. Not sure this should block this PR, though, so if you don't get to it, please add it here: https://github.com/flutter/flutter/issues/125928

reidbaker replied:
I do think that some error testing should be required before merging.

gmackall replied:
This action currently throws a CameraException. I'll add a test for it.

gmackall replied:
Added a test for 
1. if the recording is null (we haven't started a recording). In this case we throw a CameraException
2. if the videoOutputPath is null (we shouldn't generally reach this case). In this case we throw a CameraException
3. We call stop on a valid recording, and then we call it again (in this case we test everything performs normally for the first call to stop, and then we throw a CameraException on the second call).

gmackall replied:
Also added a logic and a test for calling recording twice (we silently do nothing in this case, based on a check of if the class variable `recording` is null or not).

camsim99 reviewed (approved):
Approving to unblock this. I left comments on things I think can be done later, but otherwise agree with the rest of the review comments left!

gmackall commented:
Merging currently blocked due to a test failure that I believe is unrelated and should be resolved by https://github.com/flutter/packages/pull/3947

reidbaker commented on packages/camera/camera_android_camerax/lib/src/video_capture.dart:
```diff
@@ -0,0 +1,116 @@
+  final InstanceManager instanceManager;
+
+  @override
+  void create(int identifier) {
```
🤦

gmackall commented:
To make reviewing easier (because this has been updated a lot), since your last review @reidbaker the summary of changes is:
1. Added logic in startVideoRecording to explicitely do nothing if recording is not null (there is an active recording).
2. Added a test for this case where we call startVideoRecording twice, and verify the methods we expect to be called only once are indeed called only once.
3. Made a change to the stopVideoRecording error case where the videoOutputPath is null so that we clean up the recording objects (stop the recording, set recording and pendingRecording to null) because we can't recover from this case as videoOutputPath is set in the startVideoRecording.
4. Added a test to cover the above.

camsim99 commented on packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:
```diff
@@ -421,6 +421,11 @@ class AndroidCameraCameraX extends CameraPlatform {
     assert(processCameraProvider != null);
 
+    if (recording != null) {
+      // There is currently an active recording, so do not start a new one.
```
Can you add a statement to the doc for this method that this is the expected behavior for trying to start a new recording?

gmackall replied:
Added a note that we silently return without doing anything if there is an active recording.

camsim99 commented on packages/camera/camera_android_camerax/lib/src/android_camera_camerax.dart:
```diff
@@ -444,6 +449,11 @@ class AndroidCameraCameraX extends CameraPlatform {
+      // in this error case.
+      recording!.close();
+      recording = null;
+      pendingRecording = null;
```
Same here about adding it to the doc

gmackall replied:
Added a note on the cleanup of the recording objects in the doc

camsim99 commented on packages/camera/camera_android_camerax/test/android_camera_camerax_test.dart:
```diff
@@ -561,6 +561,48 @@ void main() {
+      camera.videoOutputPath = videoOutputPath;
+
+      final XFile file = await camera.stopVideoRecording(0);
+      assert(file.path == videoOutputPath);
```
Should this be an expect statement instead of assert?

gmackall replied:
Good catch! Changed

gmackall replied:
It was actually in two places because I copied this test to make the test of "call start video recording twice", changed in both places now

camsim99 commented:
Merging manually to unblock https://github.com/flutter/packages/pull/3878
